spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/cassandra/CassandraMetricsBinder.java

```diff
@@ -0,0 +1,153 @@
+		TIMER_SNAPSHOT_METRICS.put("p98", Snapshot::get98thPercentile);
+		TIMER_SNAPSHOT_METRICS.put("p95", Snapshot::get95thPercentile);
+		TIMER_SNAPSHOT_METRICS.put("p75", Snapshot::get75thPercentile);
+		TIMER_SNAPSHOT_METRICS.put("min", Snapshot::getMin);
+		TIMER_SNAPSHOT_METRICS.put("max", Snapshot::getMax);
+		TIMER_SNAPSHOT_METRICS.put("mean", Snapshot::getMean);
+		TIMER_SNAPSHOT_METRICS.put("median", Snapshot::getMedian);
+		TIMER_SNAPSHOT_METRICS.put("stdDev", Snapshot::getStdDev);
+	}
+
+	static {
+		TIMER_METRICS.put("mean_rate", Timer::getMeanRate);
+		TIMER_METRICS.put("m1_rate", Timer::getOneMinuteRate);
+		TIMER_METRICS.put("m5_rate", Timer::getFiveMinuteRate);
+		TIMER_METRICS.put("m15_rate", Timer::getFifteenMinuteRate);
+	}
+
+	private final Iterable<Tag> tags;
+
+	private final CqlSession session;
+
+	private CassandraMetricsSettings cassandraMetricsSettings;
+
+	public CassandraMetricsBinder(@NonNull CqlSession cqlSession,
+			@NonNull CassandraMetricsSettings cassandraMetricsSettings) {
+		this.cassandraMetricsSettings = cassandraMetricsSettings;
+		this.session = cqlSession;
+		this.tags = Tags.concat(Collections.emptyList(), "sessionName", cqlSession.getName());
+	}
+
+	@Override
+	public void bindTo(@NonNull MeterRegistry meterRegistry) {
+		Assert.isTrue(this.session.getMetrics().isPresent(), "Metrics on the CqlSession needs to be present");
+
+		this.session.getMetrics().get().getRegistry().getMetrics().forEach((name, m) -> {
+			if (m instanceof Timer) {
+				registerTimer(name, (Timer) m, meterRegistry);
+			}
+			else if (m instanceof Counting) {
+				registerMeter(name, (Counting) m, meterRegistry);
+			}
+			else if (m instanceof com.codahale.metrics.Gauge) {
+				registerGauge(name, (com.codahale.metrics.Gauge) m, meterRegistry);
+			}
+		});
+	}
+
+	private void registerTimer(String metricName, Timer timer, MeterRegistry meterRegistry) {
+		String name = String.format("%s.%s", SESSION_PREFIX, metricName);
+
+		// count metric should be registered at the root level of the metric name
+		FunctionCounter.builder(name, timer, Timer::getCount).tags(this.tags).description(DESCRIPTION)
+				.register(meterRegistry);
+
+		TIMER_METRICS.forEach((n, supplier) -> {
+			String nameTimerMetric = String.format("%s.%s", name, n);
+			FunctionCounter.builder(nameTimerMetric, timer, (t) -> supplier.apply(timer).doubleValue()).tags(this.tags)
+					.baseUnit(TimeUnit.SECONDS.name()) // all rates are calculated per
+					// second
+					.description(DESCRIPTION).register(meterRegistry);
+		});
+
+		TIMER_SNAPSHOT_METRICS.forEach((n, supplier) -> {
+			String nameTimerMetric = String.format("%s.%s", name, n);
+			FunctionCounter
+					.builder(nameTimerMetric, timer,
+							(t) -> convertToTimeUnit(supplier.apply(timer.getSnapshot()).doubleValue()))
+					.tags(this.tags).baseUnit(this.cassandraMetricsSettings.getTimerTimeUnit().name())
+					.description(DESCRIPTION).register(meterRegistry);
+		});
```

Comment:
You should create the snapshot only once:

```suggestion
		Snapshot snapshot = timer.getSnapshot();
		TIMER_SNAPSHOT_METRICS.forEach((n, supplier) -> {
			String nameTimerMetric = String.format("%s.%s", name, n);
			FunctionCounter
					.builder(nameTimerMetric, timer, (t) -> convertToTimeUnit(supplier.apply(snapshot).doubleValue()))
					.tags(this.tags).baseUnit(this.cassandraMetricsSettings.getTimerTimeUnit().name())
					.description(DESCRIPTION).register(meterRegistry);
		});
```